// @bitgo/statics/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoinMap = exports.TronErc20Coin = exports.HederaToken = exports.SolCoin = exports.AvaxERC20Token = exports.AlgoCoin = exports.EosCoin = exports.StellarCoin = exports.Erc20Coin = exports.ContractAddressDefinedToken = exports.CeloCoin = exports.AccountCoin = exports.LightningCoin = exports.UtxoCoin = exports.OfcCoin = void 0;
__exportStar(require("./base"), exports);
__exportStar(require("./coins"), exports);
__exportStar(require("./networks"), exports);
__exportStar(require("./errors"), exports);
__exportStar(require("./tokenConfig"), exports);
var ofc_1 = require("./ofc");
Object.defineProperty(exports, "OfcCoin", { enumerable: true, get: function () { return ofc_1.OfcCoin; } });
var utxo_1 = require("./utxo");
Object.defineProperty(exports, "UtxoCoin", { enumerable: true, get: function () { return utxo_1.UtxoCoin; } });
var lightning_1 = require("./lightning");
Object.defineProperty(exports, "LightningCoin", { enumerable: true, get: function () { return lightning_1.LightningCoin; } });
var account_1 = require("./account");
Object.defineProperty(exports, "AccountCoin", { enumerable: true, get: function () { return account_1.AccountCoin; } });
Object.defineProperty(exports, "CeloCoin", { enumerable: true, get: function () { return account_1.CeloCoin; } });
Object.defineProperty(exports, "ContractAddressDefinedToken", { enumerable: true, get: function () { return account_1.ContractAddressDefinedToken; } });
Object.defineProperty(exports, "Erc20Coin", { enumerable: true, get: function () { return account_1.Erc20Coin; } });
Object.defineProperty(exports, "StellarCoin", { enumerable: true, get: function () { return account_1.StellarCoin; } });
Object.defineProperty(exports, "EosCoin", { enumerable: true, get: function () { return account_1.EosCoin; } });
Object.defineProperty(exports, "AlgoCoin", { enumerable: true, get: function () { return account_1.AlgoCoin; } });
Object.defineProperty(exports, "AvaxERC20Token", { enumerable: true, get: function () { return account_1.AvaxERC20Token; } });
Object.defineProperty(exports, "SolCoin", { enumerable: true, get: function () { return account_1.SolCoin; } });
Object.defineProperty(exports, "HederaToken", { enumerable: true, get: function () { return account_1.HederaToken; } });
Object.defineProperty(exports, "TronErc20Coin", { enumerable: true, get: function () { return account_1.TronErc20Coin; } });
var map_1 = require("./map");
Object.defineProperty(exports, "CoinMap", { enumerable: true, get: function () { return map_1.CoinMap; } });
