// @bitgo/sdk-coin-xlm/dist/src/lib/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.decodePrivateKey = exports.decodePublicKey = exports.encodePrivateKey = exports.encodePublicKey = exports.isValidPrivateKey = exports.isValidStellarPrv = exports.isValidPublicKey = exports.isValidStellarPub = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const stellar = __importStar(require("stellar-sdk"));
/**
 * Validates a Stellar public key
 * Stellar public keys are prefixed with 'G'
 * @param pub A Stellar public key to validate
 * @returns Whether the input is a valid Stellar public key
 */
function isValidStellarPub(pub) {
    return stellar.StrKey.isValidEd25519PublicKey(pub);
}
exports.isValidStellarPub = isValidStellarPub;
/**
 * Validates a ed25519 public key
 * @param pub A hexadecimal public key to validate
 * @returns Whether the input is a valid public key
 */
function isValidPublicKey(pub) {
    return (0, sdk_core_1.isValidEd25519PublicKey)(pub);
}
exports.isValidPublicKey = isValidPublicKey;
/**
 * Validates a Stellar private key
 * Stellar private keys are prefixed with 'S'
 * @param seed A Stellar private key to validate
 * @returns Whether the input is a valid Stellar private key
 */
function isValidStellarPrv(seed) {
    return stellar.StrKey.isValidEd25519SecretSeed(seed);
}
exports.isValidStellarPrv = isValidStellarPrv;
/**
 * Validates a ed25519 private key
 * @param prv A hexadecimal private key to validate
 * @returns Whether the input is a valid private key
 */
function isValidPrivateKey(prv) {
    return (0, sdk_core_1.isValidEd25519SecretKey)(prv);
}
exports.isValidPrivateKey = isValidPrivateKey;
/**
 * Encodes a raw public key to a G-prefixed Stellar public key
 * @param pub Raw public key
 * @returns Encoded Stellar public key
 */
function encodePublicKey(pub) {
    return stellar.StrKey.encodeEd25519PublicKey(pub);
}
exports.encodePublicKey = encodePublicKey;
/**
 * Encodes a raw private key to a S-prefixed Stellar private key
 * @param prv Raw private key
 * @returns Encoded Stellar private key
 */
function encodePrivateKey(prv) {
    return stellar.StrKey.encodeEd25519SecretSeed(prv);
}
exports.encodePrivateKey = encodePrivateKey;
/**
 * Decodes a Stellar public key to a raw public key
 * @param pub Encoded Stellar G-prefixed public key
 * @returns Raw hexadecimal public key
 */
function decodePublicKey(pub) {
    return stellar.StrKey.decodeEd25519PublicKey(pub);
}
exports.decodePublicKey = decodePublicKey;
/**
 * Decodes a Stellar private key to a raw private key
 * @param prv Encoded Stellar S-prefixed private key
 * @returns Raw hexadecimal private key
 */
function decodePrivateKey(prv) {
    return stellar.StrKey.decodeEd25519SecretSeed(prv);
}
exports.decodePrivateKey = decodePrivateKey;
