// @bitgo/smart-contracts/dist/src/eth/methods/methods.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthMethod = void 0;
const abi = __importStar(require("ethereumjs-abi"));
const ensure_1 = require("../../util/ensure");
const ethUtil = __importStar(require("ethereumjs-util"));
const ethers = __importStar(require("ethers"));
class EthMethod {
    constructor(definitions) {
        const { constant, payable, inputs, outputs, name, type } = definitions;
        this.constant = constant;
        this.payable = payable;
        this.inputs = inputs;
        this.outputs = outputs;
        this.name = name;
        this.type = type;
    }
    get id() {
        if (!this._id) {
            this._id = abi.methodID(this.name, this.inputs.map((input) => input.type)).toString('hex');
        }
        return this._id;
    }
    call(params) {
        const types = [];
        const values = [];
        let data = '';
        this.inputs.forEach((input) => {
            if (Array.isArray(input.components)) {
                let tuple = 'tuple(';
                input.components.forEach(component => {
                    (0, ensure_1.ensure)(params[component.name] !== undefined, `Missing required parameter: ${component.name}`);
                    tuple += component.type + ',';
                    values.push(params[component.name]);
                });
                tuple = tuple.slice(0, -1);
                tuple += ')';
                types.push(tuple);
                const abiElement = 'function ' + this.name + '(' + tuple + ')';
                const iface = new ethers.utils.Interface([abiElement]);
                const methodId = iface.functions[this.name].sighash;
                data = methodId + ethers.utils.defaultAbiCoder.encode(types, [values]).slice(2);
            }
            else {
                (0, ensure_1.ensure)(params[input.name] !== undefined, `Missing required parameter: ${input.name}`);
                values.push(params[input.name]);
                types.push(input.type);
                data = ethUtil.addHexPrefix(this.id + abi.rawEncode(types, values).toString('hex'));
            }
        });
        return {
            data,
            amount: '0',
        };
    }
    explain() {
        return {
            name: this.name,
            type: this.type,
            constant: this.constant,
            payable: this.payable,
            inputs: this.inputs,
            outputs: this.outputs,
        };
    }
}
exports.EthMethod = EthMethod;
