// @bitgo/smart-contracts/dist/src/base/contracts/contractInstances.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContractReader = exports.getAbiContract = exports.listContractTypes = exports.instances = exports.contracts = void 0;
const ensure_1 = require("../../util/ensure");
const contracts_1 = require("./contracts");
const abis_1 = __importDefault(require("../../../eth/abis"));
const abis_2 = __importDefault(require("../../../trx/abis"));
const instances_json_1 = __importDefault(require("../../../eth/config/instances.json"));
const instances_json_2 = __importDefault(require("../../../trx/config/instances.json"));
exports.contracts = { eth: abis_1.default, trx: abis_2.default };
exports.instances = { eth: instances_json_1.default, trx: instances_json_2.default };
/**
 * List the names of the available ABI definitions.
 * These are stored locally as JSON ABI definition files
 */
function listContractTypes(chainName) {
    return Object.keys(exports.contracts[chainName]);
}
exports.listContractTypes = listContractTypes;
/**
 * Return specific ABI
 * @param contractName The name of the contract to read the ABI for
 * @param chainName The name of the chain to read the ABI for
 * @param accessAbiValues Access values to the ABI, some contracts has sub levels to access the ABI, e.g Tron has {"entrys" : [..ABI]}
 *
 */
function getAbiContract(contractName, chainName, accessAbiValues = []) {
    let contract = exports.contracts[chainName][contractName];
    if (accessAbiValues) {
        accessAbiValues.forEach(field => {
            contract = contract[field];
        });
    }
    return (0, ensure_1.ensureExist)(contract, `Invalid JSON field`);
}
exports.getAbiContract = getAbiContract;
class ContractReader {
    constructor(chainName, defaultInstanceKey, methodsClass, methodClass) {
        if (!chainName || !defaultInstanceKey) {
            throw new Error('Default chain params not defined.');
        }
        this.chainName = chainName;
        this.defaultInstanceKey = defaultInstanceKey;
        this.methodsClass = methodsClass;
        this.methodClass = methodClass;
    }
    /**
     * Read in and parse config for instances of this contract type
     * @param contractName The name of the contract to read the config for
     * @param accessAbiValues Some contracts has sub levels to access the ABI, e.g Tron has {"entrys" : [..ABI]}
     */
    readContractInstances(contractName, accessAbiValues = []) {
        const instances = this.getInstances(contractName, this.chainName);
        const contract = this.getContract(contractName, listContractTypes(this.chainName), accessAbiValues);
        // Save them with the instance names lowercased, for easier lookup
        return this.parse(instances, contract);
    }
    /**
     * Return the contract specific ABI
     * @param contractName The name of the contract to
     * @param contractTypesList
     * @param accessAbiValues Access values to the ABI, some contracts has sub levels to access the ABI, e.g Tron has {"entrys" : [..ABI]}
     */
    getContract(contractName, contractTypesList, accessAbiValues = []) {
        (0, ensure_1.ensure)(contractTypesList.includes(contractName), `Unknown contract: ${contractName}`);
        return getAbiContract(contractName, this.chainName, accessAbiValues);
    }
    getInstances(contractName, chainName) {
        const config = exports.instances[chainName];
        (0, ensure_1.ensure)(config[contractName], `Unknown contract: ${contractName}`);
        return config[contractName];
    }
    parse(parsedConfig, contract) {
        const methodList = this.parseMethods(contract);
        const result = [];
        Object.keys(parsedConfig).forEach((instanceName) => {
            const address = parsedConfig[instanceName];
            result.push(new contracts_1.InstanceImpl(instanceName.toLowerCase(), new this.methodsClass(methodList), address));
        });
        // If no exists a default intance create one
        if (!Object.keys(parsedConfig).some((instanceName) => instanceName === 'default')) {
            result.push(new contracts_1.InstanceImpl('default', new this.methodsClass(methodList)));
        }
        return result;
    }
    parseMethods(definitions) {
        return definitions.reduce((collector, def) => {
            const method = new this.methodClass(def);
            collector.push(method);
            return collector;
        }, []);
    }
}
exports.ContractReader = ContractReader;
