// @bitgo/smart-contracts/dist/src/eth/contracts/contracts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthContract = void 0;
const methods_1 = require("../methods/methods");
const contractInstances_1 = require("../../base/contracts/contractInstances");
const ensure_1 = require("../../util/ensure");
const methods_2 = require("../../base/methods/methods");
class EthContract {
    constructor(name) {
        this.name = name;
        this.DEFAULT_INSTANCE_KEY = 'default';
        this.contractReader = new contractInstances_1.ContractReader(EthContract.chainName, this.DEFAULT_INSTANCE_KEY, methods_2.MethodsImpl, methods_1.EthMethod);
        this.contractInstances = this.contractReader.readContractInstances(name);
    }
    instance(name) {
        name = name ? name.toLowerCase() : this.DEFAULT_INSTANCE_KEY;
        const instance = this.contractInstances.find((i) => i.name === name);
        return (0, ensure_1.ensureExist)(instance, `Unknown instance: ${name}`);
    }
    listMethods() {
        return this.instance().explain().filter(method => method.type === 'function');
    }
}
exports.EthContract = EthContract;
EthContract.chainName = 'eth';
