// @bitgo/smart-contracts/dist/src/base/decoder/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parsePrimitive = exports.parseArraySubType = exports.isHex = exports.isArray = exports.formatBool = exports.Primitive = void 0;
const ethereumjs_util_1 = require("ethereumjs-util");
const ensure_1 = require("../../util/ensure");
var Primitive;
(function (Primitive) {
    Primitive[Primitive["Address"] = 0] = "Address";
    Primitive[Primitive["Bool"] = 1] = "Bool";
    Primitive[Primitive["String"] = 2] = "String";
    Primitive[Primitive["Bytes"] = 3] = "Bytes";
    Primitive[Primitive["Int"] = 4] = "Int";
})(Primitive = exports.Primitive || (exports.Primitive = {}));
function formatBool(value) {
    const intBool = (0, ethereumjs_util_1.bufferToInt)(value);
    return intBool === 1;
}
exports.formatBool = formatBool;
function isArray(type) {
    return type.includes('[]') && type.lastIndexOf(']') === type.length - 1;
}
exports.isArray = isArray;
function isHex(data) {
    return /^([0-9a-f])+$/.test(data);
}
exports.isHex = isHex;
/**
 * Take an array type and return the subtype that it is an array of
 * i.e. uint256[][] would return uint256[], and uint256[] would return uint256
 * @param type The type to parse subarray from
 * @return The subarray type
 */
function parseArraySubType(type) {
    (0, ensure_1.ensure)(isArray(type), 'Invalid type, not an array');
    return type.slice(0, type.lastIndexOf('['));
}
exports.parseArraySubType = parseArraySubType;
/**
 * Take a type string, such as uint256, and return the primitive value type that it represents
 * @param type The type to parse
 * @return The primitive value type
 * @throws if the type is unknown
 */
function parsePrimitive(type) {
    if (type.includes('address')) {
        return Primitive.Address;
    }
    if (type.includes('bool')) {
        return Primitive.Bool;
    }
    if (type.includes('string')) {
        return Primitive.String;
    }
    if (type.includes('bytes')) {
        return Primitive.Bytes;
    }
    if (type.includes('int')) {
        return Primitive.Int;
    }
    throw new Error(`Unknown type ${type}`);
}
exports.parsePrimitive = parsePrimitive;
